### PR TITLE
[WIP] Add sdxl lora support

### PR DESCRIPTION
**Pull request by StAlKeR7779**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [x] No

## Description
Add lora loading for sdxl.
NOT TESTED - I run only 2 loras, please check more(including lycoris if they already exists).

## QA Instructions, Screenshots, Recordings
https://civitai.com/models/118536/voxel-xl
image

## Added/updated tests?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by psychedelicious** [comment]
Can you link to a LoRA that should work?

File: invokeai/app/invocations/model.py [hunk 1]
```diff
@@ -262,6 +262,103 @@ def invoke(self, context: InvocationContext) -> LoraLoaderOutput:
...
 
+class SDXLLoraLoaderOutput(BaseInvocationOutput):
+    """Model loader output"""
+
+    # fmt: off
+    type: Literal["sdxl_lora_loader_output"] = "sdxl_lora_loader_output"
+
+    unet: Optional[UNetField] = Field(default=None, description="UNet submodel")
+    clip: Optional[ClipField] = Field(default=None, description="Tokenizer and text_encoder submodels")
+    clip2: Optional[ClipField] = Field(default=None, description="Tokenizer2 and text_encoder2 submodels")
```
Comment (RyanJDick) [inline invokeai/app/invocations/model.py:273]: Why are the output models `Optional`?

Comment (StAlKeR7779) [reply invokeai/app/invocations/model.py:273]: Because inputs optional too, user can input only some of this models.

**Comment by lstein** [comment]
Not working for me, but maybe I wired up the nodes wrong? I have never created a LoRA-containing graph before.

Here’s the relevant part of the node diagram:
IMG_0241
I wired the model’s unet, clip and clip2 to the corresponding LoRA inputs, and wired the outputs from the LoRA to the positive and negative compels and the text2latents. I’ve tried several SDXL LoRAs, including the one referenced above, and never saw it taking effect even with a weight of 1.0

**Comment by lstein** [comment]
I’ll go ahead and update the model probe so that the correct base model is recognized during installation.

**Comment by lstein** [comment]
Tried again this morning with a fresh graph. Still no difference between having the LoRA loaded vs not loaded. The enclosed screenshot shows a graph that uses the "Voxel" LoRA. In the gallery the upper image is with the LoRA applied, and the lower one is without. I don't see an effect even when weight is cranked high. The log shows that the lora is being loaded.

image

**Comment by lstein** [comment]
> > Tried again this morning with a fresh graph. Still no difference between having the LoRA loaded vs not loaded. The enclosed screenshot shows a graph that uses the "Voxel" LoRA. In the gallery the upper image is with the LoRA applied, and the lower one is without. I don't see an effect even when weight is cranked high. The log shows that the lora is being loaded.
> > image
> 
> In lora said that it specially trained for "voxel style" prompt, which i can't see in your prompt. Also I never use invoke as reference, I always run a1111 lora in parallel and compare outputs. //oh wait, it's sdxl topic, this one i not tested in a1111 image

I tried with a few different LoRAs and never saw a difference, but I'll try again...

Success with "voxel style":
image
Also tested with https://civitai.com/models/119405 and https://civitai.com/models/117635?modelVersionId=127510

In some cases it helps to have the trigger term in the Style part of the prompt, and in other cases it doesn't matter.

**Review by lstein — APPROVED** [review]
LGTM. Thanks for the help figuring out how to run the SDXL LoRAs.

**Review by psychedelicious — APPROVED** [review]

**Pull request merged**